src/service_control_client_impl_quota_test.cc

```diff
@@ -0,0 +1,586 @@
+
+  EXPECT_EQ(done_status, Status::OK);
+
+  // count store callback
+  EXPECT_EQ(mock_quota_transport_.on_done_vector_.size(), 0);
+
+  Statistics stat;
+  Status stat_status = client_->GetStatistics(&stat);
+
+  EXPECT_EQ(stat_status, Status::OK);
+  EXPECT_EQ(stat.total_called_quotas, 1);
+  EXPECT_EQ(stat.send_quotas_by_flush, 0);
+  EXPECT_EQ(stat.send_quotas_in_flight, 1);
+
+  EXPECT_EQ(stat.send_report_operations, 0);
+}
+
+// Cache: false, Callback: in place
+TEST_F(ServiceControlClientImplQuotaTest,
+       TestCachedQuotaWithInPlaceCallbackMultipleRequests) {
+  EXPECT_CALL(mock_quota_transport_, Quota(_, _, _))
+      .WillOnce(Invoke(&mock_quota_transport_,
+                       &MockQuotaTransport::AllocateQuotaWithInplaceCallback));
+
+  Status done_status = Status::UNKNOWN;
+  AllocateQuotaResponse quota_response;
+
+  client_->Quota(quota_request1_, &quota_response,
+                 [&done_status](Status status) { done_status = status; });
+
+  EXPECT_EQ(done_status, Status::OK);
+
+  // call Quota 10 times
+  for (int i = 0; i < 10; i++) {
+    client_->Quota(quota_request1_, &quota_response,
+                   [&done_status](Status status) { done_status = status; });
+    EXPECT_EQ(done_status, Status::OK);
+  }
+
+  // count store callback
+  EXPECT_EQ(mock_quota_transport_.on_done_vector_.size(), 0);
+
+  Statistics stat;
+  Status stat_status = client_->GetStatistics(&stat);
+
+  EXPECT_EQ(stat_status, Status::OK);
+  EXPECT_EQ(stat.total_called_quotas, 11);
+  EXPECT_EQ(stat.send_quotas_by_flush, 0);
+  EXPECT_EQ(stat.send_quotas_in_flight, 1);
+
+  EXPECT_EQ(stat.send_report_operations, 0);
+}
+
+// Cache: false, Callback: local in place
+TEST_F(ServiceControlClientImplQuotaTest,
+       TestNonCachedQuotaWithLocalInPlaceCallback) {
+  Status done_status = Status::UNKNOWN;
+  AllocateQuotaResponse quota_response;
+
+  int callbackExecuteCount = 0;
+
+  client_->Quota(
+      quota_request1_, &quota_response,
+      [&done_status](Status status) { done_status = status; },
+      [this, &callbackExecuteCount](const AllocateQuotaRequest& request,
+                                    AllocateQuotaResponse* response,
+                                    TransportDoneFunc on_done) {
+        callbackExecuteCount++;
+        on_done(Status::OK);
+      });
+
+  EXPECT_EQ(done_status, Status::OK);
+
+  EXPECT_EQ(callbackExecuteCount, 1);
+
+  // count store callback
+  EXPECT_EQ(mock_quota_transport_.on_done_vector_.size(), 0);
+
+  Statistics stat;
+  Status stat_status = client_->GetStatistics(&stat);
+
+  EXPECT_EQ(stat_status, Status::OK);
+  EXPECT_EQ(stat.total_called_quotas, 1);
+  EXPECT_EQ(stat.send_quotas_by_flush, 0);
+  EXPECT_EQ(stat.send_quotas_in_flight, 1);
+
+  EXPECT_EQ(stat.send_report_operations, 0);
+}
+
+// Cache: false, Callback: in place
+TEST_F(ServiceControlClientImplQuotaTest,
+       TestNonCachedQuotaWithInPlaceCallbackInSync) {
```

Comment:
The first request was not cached which is same as cache disabled.  InternalTestNonCachedCheckWithStoredCallback is using the same logic to test non cached test.

https://github.com/cloudendpoints/service-control-client-cxx/blob/master/src/service_control_client_impl_test.cc#L462